src/Client.js

```diff
@@ -0,0 +1,114 @@
+import Api from './core/Api';
+import SocketHandler from './core/SocketHandler';
+import { Event } from './core/ChainEvent';
+export default class Client {
+  constructor(
+    apiUrl = 'http://localhost',
+    apiPort = '4000',
+    socketUrl = 'ws://127.1:4000/socket'
+  ) {
+    this._api = new Api(apiUrl, apiPort);
+    this._socket = new SocketHandler(socketUrl);
+  }
+
+  async init() {
+    await this._socket.init();
+    await this.once('api', Event.API_JOIN);
+  }
+
+  api() {
+    return this._api;
+  }
+
+  socket() {
+    return this._socket;
+  }
+
+  channel(id) {
+    return this._socket.channel(id);
+  }
+
+  channels() {
+    return this._socket._channels;
+  }
+
+  stream(id) {
+    return this.channel(id).stream();
+  }
+
+  once(name, event) {
+    return this.channel(name).once(event);
+  }
+
+  async create(options) {
+    this.channel('api').push('start', { ...options });
+
+    const {
+      payload: {
+        response: { id }
+      }
+    } = await this.once('api', Event.CHAIN_CREATED);
+
+    if (options.step_id) {
+      return await this.sequenceEvents(id, [
+        Event.CHAIN_STARTED,
+        Event.CHAIN_DEPLOYING,
+        Event.CHAIN_STATUS_ACTIVE,
+        Event.CHAIN_DEPLOYED,
+        Event.CHAIN_READY
+      ]);
+    } else {
+      return await this.sequenceEvents(id, [
+        Event.CHAIN_STARTED,
+        Event.CHAIN_STATUS_ACTIVE,
+        Event.CHAIN_READY
+      ]);
+    }
+  }
+
+  stop(id) {
+    this.channel(id).push('stop');
+  }
+
+  restart(id) {
+    this.channel('api').push('start_existing', { id });
+  }
+
+  async delete(id) {
+    const { details } = await this.api().getChain(id);
+
+    this.stop(id);
+    await this.once(id, 'stopped');
+    if (!details.config.clean_on_stop) {
+      this.api().deleteChain(id);
+    }
+    await this.socket()._sleep(2000); // FIXME: Have to wait for server to update
+  }
+
+  takeSnapshot(id, description = '') {
+    this.channel(id).push('take_snapshot', { description });
+  }
+
+  restoreSnapshot(id, snapshot) {
+    this.channel(id).push('revert_snapshot', { snapshot });
+  }
+
+  async sequenceEvents(id, eventNames) {
+    // - Takes an array of event constants
+    // - Creates an object of promises, keyed by their names
+    // - Waits for them all to resolve
+    const objPromise = {};
+    for (const event of eventNames) {
```

Comment:
we can make this a bit more functional for ex.:

```
const objPromise = eventNames.reduce(...)
```

but thinking more about this, can't the same result be achieved by doing
```
await Promise.all[eventNames.map(ev => this.once(id, ev)]
```

Not sure I get the sequence aspects here?

Reply:
I had initially gone with the Promise.all but that returned an array. I'll check it out.

I did run into a bit of trouble with the sequencing. For a `start` push, we would expect a `started` -> `status_changed` -> `ready` event response sequence. When the  `status_changed` would fire, the `ready` would fire directly. If I'm awaiting on these events in a loop, by the time the `status_changed` promise would resolve, loop to await on the `ready` promise, the `ready` event would have been fired and not caught.

That's why I'm setting up the promises beforehand, and then awaiting on them in their expected order.